packages/useTrie/src/index.ts

```diff
@@ -55,21 +57,22 @@ class Trie implements ITrie {
     words.forEach(word => this.add(word, this.getText));
   }
 
-  private normalizeWord = (word: Word) =>
-    this.isCaseInsensitive
+  private normalizeWord(word: Word): string {
+    return this.isCaseInsensitive
       ? (typeof word === 'string'
           ? word
           : this.getText(word) || ''
         ).toLowerCase()
       : typeof word === 'string'
       ? word
       : this.getText(word);
+  }
 
   /*
    * @param {string} word A word to check if it exists in the trie
    * @param {boolean} exactSearch Return true only if the exact word is stored
    */
-  has = (wordToSearch: string, exactSearch: boolean = true): boolean => {
+  has(wordToSearch: string, exactSearch: boolean = true): boolean {
```

Comment:
I don't know why but `uglifyjs` is not keeping the `has` method when the arrow method is not used... :thinking: 
I had to revert it back b9f1aec40ef78ca40b7ebd53eaba73f663f63e43 to build a minified version...

Reply:
v1.0.8 has that change.
https://www.npmjs.com/package/@cshooks/usetrie

Reply:
Weird. Maybe look at terser. See this thread https://twitter.com/dan_abramov/status/1060245441084825600 or just google terser.js

Reply:
It is so funny I was looking at terser just now :smile: 

I've used it once and forgot about it.
https://github.com/dance2die/calendar-dates/blob/master/rollup.config.js#L26

Reply:
From what I've read, https://iamturns.com/typescript-babel/
Babel Typescript is not perfect but as my code base isn't using any of new feature, I am thinking about going with Webpack || Rollup (to transpile as ES6 -> terser/minify -> babel to ES5)

Would it be too much for Webpack/Roll up? or should I keep it as arrow methods?

Reply:
:thinking: I can run builds in a sequence so I can actually pull that sequence in package.json... Maybe I will try that first :smile:

Reply:
That sequence (tsc -> es6 -> terser -> babel -> minify) still adds the non-arrow member method to the prototype(`Trie.prototype.has`), not as an instance (`this.has`) :thinking: 
I will stop for now & off to :bed:

Reply:
TypeScript [transpiles](https://www.typescriptlang.org/play/index.html#src=class%20Trie%20%7B%0D%0A%20%20%20%20has(word%3A%20string)%3A%20boolean%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20true%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20Trie2%20%7B%0D%0A%20%20%20%20has%20%3D%20(word%3A%20string)%3A%20boolean%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20true%3B%0D%0A%20%20%20%20%7D%0D%0A%7D) the following code

```typescript
class Trie {
    has(word: string): boolean {
        return true;
    }
}

class Trie2 {
    has = (word: string): boolean => {
        return true;
    }
}
```
as

```js
var Trie = /** @class */ (function () {
    function Trie() {
    }
    Trie.prototype.has = function (word) {
        return true;
    };
    return Trie;
}());
var Trie2 = /** @class */ (function () {
    function Trie2() {
        this.has = function (word) {
            return true;
        };
    }
    return Trie2;
}());

```

And works well.
![image](https://user-images.githubusercontent.com/8465237/55047384-f6b5eb80-501a-11e9-923b-5e19d6f0b9d2.png)

This means, I'd have to play around with the minify/mangle configurations :smile: 

Task for tomorrow

Reply:
It's OK to have it on the prototype. You want to prefer that because methods on the prototype are shared between instances. What's the issue you have when it's on the prototype?

Reply:
`tsc` compiles correctly using prototypes  (`has/add/remove` etc as prototypes).

```js
var React = __importStar(require("react")),
    TrieNode = function TrieNode(e) {
        void 0 === e && (e = ""), this.character = e, this.next = {}
    },
    Trie = function () {
        function Trie(e, t, r) {
            void 0 === e && (e = []), void 0 === t && (t = !0), void 0 === r && (r = function (e) {
                return e
            }), this.isCaseInsensitive = t, this.getText = r, this.root = new TrieNode(""), this.isCaseInsensitive = t, this.buildTrie(e)
        }
        return Trie.prototype.buildTrie = function (e) {
            var t = this;
            e.forEach(function (e) {
                return t.add(e, t.getText)
            })
        }, Trie.prototype.normalizeWord = function (e) {
            return this.isCaseInsensitive ? ("string" == typeof e ? e : this.getText(e) || "").toLowerCase() : "string" == typeof e ? e : this.getText(e)
        }, Trie.prototype.has = function (e, t) {
            void 0 === t && (t = !0);
            var r = this.normalizeWord(e);
            if ("" === r) return !1;
            for (var i = this.root, n = 0; n < r.length; n++) {
                var o = r[n];
                if (!i.next[o]) return !1;
                i = i.next[o]
            }
            return !t || !!i.id
        }, Trie.prototype.add = function (e, t) {...
```

But  uglifyjs is turning it into like this.

```js
var React = __importStar(require("react")),
  TrieNode = function a() {
    var b = 0 < arguments.length && arguments[0] !== void 0 ? arguments[0] : "";
    _classCallCheck(this, a), (this.character = b), (this.next = {});
  },
  Trie = /*#__PURE__*/ (function() {
    function a() {
      var b =
          0 < arguments.length && void 0 !== arguments[0] ? arguments[0] : [],
        c = !(1 < arguments.length && void 0 !== arguments[1]) || arguments[1],
        d =
          2 < arguments.length && void 0 !== arguments[2]
            ? arguments[2]
            : function(a) {
                return a;
              };
      _classCallCheck(this, a),
        (this.isCaseInsensitive = c),
        (this.getText = d),
        (this.root = new TrieNode("")),
        (this.isCaseInsensitive = c),
        this.buildTrie(b);
    }
    return (
      _createClass(a, [
        {
          key: "buildTrie",
          value: function buildTrie(a) {
            var b = this;
            a.forEach(function(a) {
              return b.add(a, b.getText);
            });
          }
        },
        {
          key: "normalizeWord",
          value: function normalizeWord(a) {
            return this.isCaseInsensitive
              ? ("string" == typeof a ? a : this.getText(a) || "").toLowerCase()
              : "string" == typeof a
              ? a
              : this.getText(a);
          }
        },
        {
          key: "has",
          value: function has(a) {
            var b =
                !(1 < arguments.length && void 0 !== arguments[1]) ||
                arguments[1],
              c = this.normalizeWord(a);
            if ("" === c) return !1;
            for (var d, e = this.root, f = 0; f < c.length; f++) {
              if (((d = c[f]), !e.next[d])) return !1;
              e = e.next[d];
            }
            return !b || !!e.id;
          }
        },
```
where `__createClass` looks like following

```js
function _createClass(a, b, c) {
  return (
    b && _defineProperties(a.prototype, b), c && _defineProperties(a, c), a
  );
}
```

I didn't include the full source above as I am still playing around with the uglifyjs...

Reply:
It seems like it's the `-c or ----consolidate-primitive-values` flag converting the prototype into using define properties.

https://github.com/mishoo/UglifyJS#usage

Reply:
I seemed to be at the wrong spot as it looks like `useTrie` is returning a trie containing `add/remove` only

![image](https://user-images.githubusercontent.com/8465237/55120827-e3b02380-50cd-11e9-96b5-0d6b1c30e877.png)

Reply:
Reducer looked OK though :thinking: 

```js
function reducer(e, t) {
  switch (t.type) {
    case 'ADD':
      return (
        e.trie.add(t.payload.word),
        __assign({}, e, { trie: e.trie })
      );
    case 'REMOVE':
      return (
        e.trie.remove(t.payload.word),
        __assign({}, e, { trie: e.trie })
      );
    default:
      return e;
  }
}
function useTrie(e, t, r) {
  void 0 === t && (t = !0),
    void 0 === r &&
      (r = function(e) {
        return e;
      });
  var i = new Trie(e, t, r),
    n = __read(React.useReducer(reducer, { trie: i, word: '' }), 2),
    o = n[0],
    s = n[1];
  return __assign({}, o.trie, {
    add: function add(e) {
      s({ type: 'ADD', payload: { trie: i, word: e } });
    },
    remove: function remove(e) {
      s({ type: 'REMOVE', payload: { trie: i, word: e } });
    },
  });
}
(exports.Trie = Trie), (exports.default = useTrie);
```

Reply:
Wait a sec... Doesn't spread work on prototypes? Let me see...

Reply:
Ah ha!

Spreading in reducer didn't copy `prototype.has`.
![prototype doesn't spread](https://user-images.githubusercontent.com/8465237/55121072-1c9cc800-50cf-11e9-9c0f-d9eef74edf1c.png)

Reply:
Spread works only on own & enumerable properties.
https://dmitripavlutin.com/object-rest-spread-properties-javascript/#12ownproperties

And MDN says `propertype` is not enumerable.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/prototype

I gotta blame my implementation of mutable Trie object instance & try to return a new object instance in `useTrie`...

Reply:
I "might" convert the Trie class as immutable data structure.

This seems to help (requires Egghead subscription)
https://egghead.io/lessons/javascript-avoiding-mutations-in-javascript-with-immutable-data-structures